### PR TITLE
Torchlight

### PR DESCRIPTION
**Pull request by Prakkkmak**

### CODE REVIEW

**Comment by Prakkkmak** [comment]
[Torch Light](https://trello.com/c/MXjWloVI/16-torch-light)

File: code/Modules/Items/Carriables/ItemTorchlight.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        private const float Falloff = 1.0f;
+        private const float LinearAttenuation = 0.0f;
+        private const float QuadraticAttenuation = 1.0f;
+        private const int Brightness = 2;
+        private SpotLightEntity worldLight;
+        private SpotLightEntity viewLight;
+        private const int InnerConeAngle = 20;
+        private const int OuterConeAngle = 40;
+        private const float FogStength = 1.0f;
+        public ItemTorchlight()
```
Comment (Nowp) [inline code/Modules/Items/Carriables/ItemTorchlight.cs:18]:
```suggestion
        private const float FogStength = 1.0f;
        
        public ItemTorchlight()
```

File: code/Modules/Items/Carriables/ItemTorchlight.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        public ItemTorchlight(ItemData data) : base(data)
+        {
+        }
+        
+        //TODO Replace with the NeckCamera 
+        public override string ViewModelPath => "weapons/rust_flashlight/v_rust_flashlight.vmdl";
+        
+        private static Vector3 LightOffset => Vector3.Forward * 10;
+        [Net, Local, Predicted]
+        private bool LightEnabled { get; set; } = true;
```
Comment (Nowp) [inline code/Modules/Items/Carriables/ItemTorchlight.cs:31]: Why Local ? Other players need to see if the torchlight is on 
We'll need to test this

File: code/Modules/Items/Carriables/ItemTorchlight.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+        
+
+        private TimeSince timeSinceLightToggled;
```
Comment (Nowp) [inline code/Modules/Items/Carriables/ItemTorchlight.cs:35]:
```suggestion

        private TimeSince timeSinceLightToggled;
```

File: code/Modules/Items/Carriables/ItemTorchlight.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                }
+                else
+                {
+                    Deactivate();
+                }
+            }
+        }
+
+    }
+}
```
Comment (Nowp) [inline code/Modules/Items/Carriables/ItemTorchlight.cs:152]: Copy paste of S&box, spaces within parentheses need to be removed

File: code/Modules/Items/Carriables/ItemTorchlight.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        private const int Range = 512;
+        private const float Falloff = 1.0f;
+        private const float LinearAttenuation = 0.0f;
+        private const float QuadraticAttenuation = 1.0f;
+        private const int Brightness = 2;
+        private SpotLightEntity worldLight;
+        private SpotLightEntity viewLight;
+        private const int InnerConeAngle = 20;
+        private const int OuterConeAngle = 40;
+        private const float FogStength = 1.0f;
```
Comment (Nowp) [inline code/Modules/Items/Carriables/ItemTorchlight.cs:17]:
```suggestion
        private const int Range = 512;
        private const float Falloff = 1.0f;
        private const float LinearAttenuation = 0.0f;
        private const float QuadraticAttenuation = 1.0f;
        private const int Brightness = 2;
        private const int InnerConeAngle = 20;
        private const int OuterConeAngle = 40;
        private const float FogStength = 1.0f;
        private SpotLightEntity worldLight;
        private SpotLightEntity viewLight;
```

File: code/Modules/Items/Carriables/ItemTorchlight.cs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        
+
+        private TimeSince timeSinceLightToggled;
+        
+        public override void Spawn()
+        {
+            base.Spawn();
+
+            worldLight = CreateLight();
+            worldLight.SetParent( this, "slide", new Transform( LightOffset ) );
```
Comment (Nowp) [inline code/Modules/Items/Carriables/ItemTorchlight.cs:42]: Magic number and spaces within parenthesis

File: code/Modules/Items/Carriables/ItemTorchlight.cs [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+            worldLight.EnableHideInFirstPerson = true;
+            worldLight.Enabled = false;
+        }
+        
+        public override void CreateViewModel()
+        {
+            base.CreateViewModel();
+
+            viewLight = CreateLight();
+            viewLight.SetParent( ViewModelEntity, "light", new Transform( LightOffset ) );
```
Comment (Nowp) [inline code/Modules/Items/Carriables/ItemTorchlight.cs:52]: same

**Review by Nowp — CHANGES_REQUESTED** [review]

File: code/Modules/Items/Carriables/ItemTorchlight.cs [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+            base.Simulate(cl);
+
+            bool toggle = Input.Pressed(InputButton.Flashlight) || Input.Pressed(InputButton.Attack1);
+
+            if (timeSinceLightToggled > 0.1f && toggle)
+            {
+                LightEnabled = !LightEnabled;
+
+                PlaySound(LightEnabled ? "flashlight-on" : "flashlight-off");
```
Comment (Nowp) [inline code/Modules/Items/Carriables/ItemTorchlight.cs:92]: Magic number

File: code/Modules/Items/Carriables/ItemTorchlight.cs [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        private const float LinearAttenuation = 0.0f;
+        private const float QuadraticAttenuation = 1.0f;
+        private const int Brightness = 2;
+        private const int InnerConeAngle = 20;
+        private const int OuterConeAngle = 40;
+        private const float FogStength = 1.0f;
+        private const string SlideAttachementName = "slide";
+        private const string LightAttachementName = "light";
+        private const string FlashLighOnName = "flashlight-on";
+        private const string FlashLighOffName = "flashlight-off";
```
Comment (Nowp) [inline code/Modules/Items/Carriables/ItemTorchlight.cs:19]: Missing "t" and we should know that this is a sound name
(I cannot add suggestions i'm in the train)

**Review by Nowp — APPROVED** [review]

**Pull request merged**